src/reducers/comments.tsx

```diff
@@ -112,12 +112,17 @@ export const createEmptyCommentInfo = (): CommentInfo => {
 };
 
 export type CommentKeyParams = {
+  commentId: number | undefined;
```

Comment:
It's surprising to me that you would need to add `commentId` to the key params. The original goal of `CommentKeyParams` was to support adding new comments that don't yet exist -- it creates a key for that data since we don't yet have a real comment ID.

For editing a comment, you know it exists, you have a real comment ID, and could put data on the existing comment object. Have you considered creating a new action (like `beginEditComment({ id: ... })`) and having the reducer simply add a property like `isEditing` to [`Comment`](https://github.com/mozilla/addons-code-manager/blob/9835897948926e3115210043fdadd4a1565e670f/src/reducers/comments.tsx#L43-L52)? 

You may want to model the flow on deletion since that uses props like `comment.beginDelete` and it deals with a similar case where the comment object already exists.

If that approach works, maybe it would be a simpler patch?

Reply:
I tried this before and found that will introduce a lot of duplication, since `savingComment`, `considerDiscard`, `pendingCommentText` (already in[ `CommentInfo`](https://github.com/mozilla/addons-code-manager/blob/master/src/reducers/comments.tsx#L70)) need to be added into `Comment` as well, and respective actions to maintain these properties. That is why I chose to reuse `CommentInfo` which is originally for new comment, to track the status of comments in editing. Without adding `commentId` to the key params, we can only keep the status of one comment in editing (or new comment) for one line, I guess we do not want that restriction.

Reply:
> I tried this before and found that will introduce a lot of duplication, since `savingComment`, `considerDiscard`, `pendingCommentText` (already in[ `CommentInfo`](https://github.com/mozilla/addons-code-manager/blob/master/src/reducers/comments.tsx#L70)) need to be added into `Comment` as well

Why do they need to be duplicated?

Can't the edit button just dispatch a `beginEditing({ id: ... })` action so `CommentList` can look for `comment.beginEdit` and set `readOnly` accordingly?

Here is an example of what I mean patched against master but I haven't tried going further with it:

```diff
diff --git a/src/components/CommentList/index.tsx b/src/components/CommentList/index.tsx
index c5e5525..c7680a0 100644
--- a/src/components/CommentList/index.tsx
+++ b/src/components/CommentList/index.tsx
@@ -93,7 +93,12 @@ export class CommentListBase extends React.Component<Props> {
         );
       }
       for (const id of commentInfo.commentIds) {
-        comments.push(<Comment {...base} commentId={id} key={id} readOnly />);
+        // This would actually be set by looking for
+        // state.byId[id].beginEdit or something like that.
+        const editing = true;
+        comments.push(
+          <Comment {...base} commentId={id} key={id} readOnly={!editing} />,
+        );
       }
     }
 
```

Reply:
Do we want to support editing multiple comments in the same line as shown bellow? Duplication of `savingComment` and other keys (or adding `commentId` into key params) is to track the states (saving/considerDiscard) of multiple comments in the same line.

![image](https://user-images.githubusercontent.com/4984681/69177467-d284ae80-0ac4-11ea-8262-2abd0af41b9e.png)

Reply:
I think about only supporting one comment in editing for each line. It should be possible by only adding `beginEdit`. But I am not sure if we want this (preventing the user from editing other comments in the same line when one comment is being edited).

In my patch, `comments.byKey.[commentId specified].beginNewComment` is playing the same role as `beginEdit` you mentioned, even if it looks not nice...

Reply:
Ah, yes, I see now! Thanks, I commented more about it in https://github.com/mozilla/addons-code-manager/pull/1241#discussion_r348147622 so I will resolve this thread.